cmd/run.go

```diff
@@ -105,17 +113,95 @@ func (c *cmdRun) run(cmd *cobra.Command, args []string) (err error) {
 		return err
 	}
 
-	// TODO: create a MetricsEngine here and add its ingester to the list of
-	// outputs (unless both NoThresholds and NoSummary were enabled)
+	executionState := execScheduler.GetState()
+	metricsEngine, err := engine.NewMetricsEngine(executionState)
+	if err != nil {
+		return err
+	}
+	if !testRunState.RuntimeOptions.NoSummary.Bool || !testRunState.RuntimeOptions.NoThresholds.Bool {
+		// We'll need to pipe metrics to the MetricsEngine if either the
+		// thresholds or the end-of-test summary are enabled.
+		outputs = append(outputs, metricsEngine.CreateIngester())
+	}
+
+	if !testRunState.RuntimeOptions.NoSummary.Bool {
+		defer func() {
+			logger.Debug("Generating the end-of-test summary...")
+			summaryResult, hsErr := test.initRunner.HandleSummary(globalCtx, &lib.Summary{
+				Metrics:         metricsEngine.ObservedMetrics,
+				RootGroup:       testRunState.Runner.GetDefaultGroup(),
+				TestRunDuration: executionState.GetCurrentTestRunDuration(),
+				NoColor:         c.gs.flags.noColor,
+				UIState: lib.UIState{
+					IsStdOutTTY: c.gs.stdOut.isTTY,
+					IsStdErrTTY: c.gs.stdErr.isTTY,
+				},
+			})
+			if hsErr == nil {
+				hsErr = handleSummaryResult(c.gs.fs, c.gs.stdOut, c.gs.stdErr, summaryResult)
+			}
+			if hsErr != nil {
+				logger.WithError(hsErr).Error("failed to handle the end-of-test summary")
+			}
+		}()
+	}
 
-	// TODO: remove this completely
-	// Create the engine.
-	initBar.Modify(pb.WithConstProgress(0, "Init engine"))
-	engine, err := core.NewEngine(testRunState, execScheduler, outputs)
+	// Create and start the outputs. We do it quite early to get any output URLs
+	// or other details below. It also allows us to ensure when they have
+	// flushed their samples and when they have stopped in the defer statements.
+	initBar.Modify(pb.WithConstProgress(0, "Starting outputs"))
+	outputManager := output.NewManager(outputs, logger, func(err error) {
+		if err != nil {
+			logger.WithError(err).Error("Received error to stop from output")
+		}
+		// TODO: attach run status and exit code?
+		runAbort(err)
+	})
+	samples := make(chan metrics.SampleContainer, test.derivedConfig.MetricSamplesBufferSize.Int64)
```

Comment:
It impacts a lot of parts, so I'm not sure. Could it now be directly a field of the OutputManager, instead of having it injected? Maybe it helps also to centralize control over it (the stop operation), and we can just pass it to the execution scheduler.

Reply:
Hmm maybe, though I'd probably prefer this to be done in a separate PR :thinking: it's not very straightforward because of the thresholds, which need the sequence from https://github.com/grafana/k6/pull/2815#discussion_r1085273225, but the `OutputManager` doesn't know that the `MetricsEngine`'s `Ingester` is in any way a special output... 

With this PR, `samples` is created here, then it is passed around wherever it is needed, and then finally it is closed in a `defer` further down this same function. That is a relatively clean and Go-idiomatic bit of code and already a massive improvement to the previous state :sweat_smile: So any further improvements can probably wait for things to settle down a bit.